core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -202,6 +226,18 @@ public long acquireExecutionMemory(long required, MemoryConsumer requestingConsu
         logger.debug("Task {} acquired {} for {}", taskAttemptId, Utils.bytesToString(got),
           requestingConsumer);
       }
+
+      if (mode == MemoryMode.OFF_HEAP) {
+        synchronized (offHeapMemoryLock) {
```

Comment:
The whole function `acquireExecutionMemory` is under the protection of `synchronized (this)`, and the release memory can be got from `trySpillAndAcquire()`:

```scala
long released = consumerToSpill.spill(requested, requestingConsumer);
```

 So I don't think we need extra lock here.

Reply:
`acquireExecutionMemory` is synchronized but `releaseExecutionMemory` is not synchronized.

While we maintain current memory in both places, we can either
- `synchronized (this)` on `releaseExecutionMemory`
- or add another lock, for smaller lock granularity

Reply:
Got it.

Actually, could we calculate `consumers.map(.used).sum + got` as the peak memory at the end of `acquireExecutionMemory`?

Reply:
I think that's doable. That way we don't even need to maintain the current memory, instead we update the peak memory after each `acquireExecutionMemory` call.

Updated the code, please take another look